CKA_ALLOWED_MECHANISMS - it should be a int_seq

vjardin opened the pull request:
memtype was not properly set.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
CKA_ALLOWED_MECHANISMS - it should be a int_seq

memtype was not properly set.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin commented:
better option in order to fix https://github.com/tpm2-software/tpm2-pkcs11/issues/805

vjardin commented:
then, side question @williamcroberts > since the YAML content of the SQL database has to be cleaned/updated in order to convert the str: to a seq: for a production system, what would be the best option in order to do so ?

williamcroberts commented:
> then, side question @williamcroberts > since the YAML content of the SQL database has to be cleaned/updated in order to convert the str: to a seq: for a production system, what would be the best option in order to do so ?

We will need to bump the database number and add a handler to fix this.

An example of a handler is:
  - https://github.com/tpm2-software/tpm2-pkcs11/blob/master/src/lib/db.c#L1586

This handler would look at the YAML contents for each object and check its CKA_ALLOWED_MECHASMS is a sequence and not a hex string, if it is, it'll need to parse it apart and update it to the sequence list by calling the YAML emitter again.

Then you add the handler to this list at the end:
  - https://github.com/tpm2-software/tpm2-pkcs11/blob/master/src/lib/db.c#L2201

Then you bump the db version number:
  - https://github.com/tpm2-software/tpm2-pkcs11/blob/master/src/lib/db.c#L42

vjardin commented:
thanks @williamcroberts , it'll be checked.

And regarding the current fix, does it look good ?

vjardin commented:
Regarding the marshaling and the rewrite of the yaml content, I am not an expert with the libyaml, do you have a tip for doing it in a smart way ?

williamcroberts commented:
> Regarding the marshaling and the rewrite of the yaml content, I am not an expert with the libyaml, do you have a tip for doing it in a smart way ?

The attributes should be right in the "tobject", so I would just call the emitter and replace their contents. You could probably call `_db_update_tobject_attrs(db, tobj->id, tobs->attrs)`;

To detect if they need updating, I would call:
```C
for each tobject:
    CK_ATTRIBUTE_PTR a = attr_get_attribute_by_type(tobj->attrs, CKA_ALLOWED_MECHANISMS);
    CK_BYTE type = type_from_ptr(a->pValue, a->ulValueLen);
    if (type != TYPE_BYTE_INT_SEQ) {
        _db_update_tobject_attrs(db, tobj->id, tobs->attrs);
    }
```

vjardin committed:
Enforce new version database 8 - no yaml bug

From version 8+ the CKA_ALLOWED_MECHANISMS will
be properly stored using yaml seq. In case they
are not, a hot patch of the database is performed.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin commented:
@williamcroberts > is it the proper marshalling ?
  https://github.com/tpm2-software/tpm2-pkcs11/pull/808/commits/ef1547b62ce70b394b5a74aa15567abf9c445463
  
please, do not merge it yet, it is a draft code for sharing my understanding based on your comments.

vjardin commented:
I do not understand how to get the sonar report from these logs:
```
scan-build: Run 'scan-view /tmp/scan-build-2023-01-10-161330-53449-1' to examine bug reports.
Error: Process completed with exit code 1.
```
 ??

williamcroberts commented:
> @williamcroberts > is it the proper marshalling ? [ef1547b](https://github.com/tpm2-software/tpm2-pkcs11/commit/ef1547b62ce70b394b5a74aa15567abf9c445463)
> 
> please, do not merge it yet, it is a draft code for sharing my understanding based on your comments.

looks about right, i'd have to dig into why its failing to give you better feedback.

vjardin commented:
> looks about right, i'd have to dig into why its failing to give you better feedback.

Can we run the test with more verbosity in order to understand the root cause of the failures ?

williamcroberts commented:
> > looks about right, i'd have to dig into why its failing to give you better feedback.
> 
> Can we run the test with more verbosity in order to understand the root cause of the failures ?

It tells you if select the failure tab:
```
+ tpm2_ptool addcert --label=label --key-label=ec1 --path=/tmp/tpm_simulator_BBsIAP /tmp/tpm_simulator_BBsIAP/cert.pem.ec1
DB Upgrade failed: "DB Version exceeds library version: 8 > 7"
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
```

I forgot the python code also needs the same logic added to it.

See the logs
https://github.com/tpm2-software/tpm2-pkcs11/actions/runs/3885289344/jobs/6628978268#step:4:365

vjardin committed:
Convert CKA_ALLOWED_MECHANISMS 32 bits

From version 8+ the CKA_ALLOWED_MECHANISMS will
be properly stored using yaml seq. In case they
are not, a hot patch of the database is performed.
Python version of the db.c code.

Signed-off-by: Vincent Jardin <vjardin@free.fr>
Suggested-by: Adrien Gausseran <adrien.gausseran@ekinops.com>

vjardin commented:
> I forgot the python code also needs the same logic added to it.

thanks for the tip, python version push'd, so far so good for the CI ;)

williamcroberts commented on tools/tpm2_pkcs11/db.py:
```diff
@@ -532,6 +532,23 @@ def _update_on_7(self, dbbakcon):
+        tobjs = c.fetchall()
+
+        for t in tobjs:
+            attrs = yaml.safe_load(io.StringIO(t['attrs']))
```
does this actually work, if you had a V7 database? As t['attrs'] is either a list or a string, shouldn't the check be:
1. if string, break it apart based on sizeof the data type? You could use something `ctypes.sizeof(c_ulong)`, then take that result multiply it by two and split the hex strings on that and create a list, then convert that list to yaml update the attributes and then update the DB

vjardin replied:
right, it was wrong ! hold on.

vjardin replied:
@williamcroberts , is https://github.com/tpm2-software/tpm2-pkcs11/commit/f15726cf91de0d5f4612a957e981981b020f1835 what you'd like ?

williamcroberts replied:
> @williamcroberts , is [f15726c](https://github.com/tpm2-software/tpm2-pkcs11/commit/f15726cf91de0d5f4612a957e981981b020f1835) what you'd like ?

Yeah that looks right, just some minor nits on that patch

vjardinMisc replied:
Which ones ? Which modifications would you like?

vjardinMisc commented:
Hi @williamcroberts , I think it is all done. If any other requests, feel free.

williamcroberts commented:
opensuse-leap CI is having issues, but I think it's caused from an update to the images and not this patch, so merging.

williamcroberts commented on src/lib/db.c:
```diff
@@ -2131,6 +2132,69 @@ static CK_RV dbup_handler_from_6_to_7(sqlite3 *updb) {
+        /* for each tobject */
+        CK_ATTRIBUTE_PTR a = attr_get_attribute_by_type(tobj->attrs, CKA_ALLOWED_MECHANISMS);
+        CK_BYTE type = type_from_ptr(a->pValue, a->ulValueLen);
+        if (type != TYPE_BYTE_INT_SEQ) {
```
I think we needed to change the data type here to TYPE_BYTE_INT_SEQ so the emitter updates everything, so I think this didn't actually do anything.

williamcroberts replied:
I have a PR for this fix: https://github.com/tpm2-software/tpm2-pkcs11/pull/924 which also includes the python code fix from the discussion in #805
